### ISSUE TITLE
Array reassignment doesn't blow up LTL and LTEL

### ISSUE THREAD

**Issue by kelloggm** [kelloggm/checker-framework#30]
I've checked in a test case for this: Reassignment.java. 77304fffeddb16cd4d8cd8f18664b896360d5747

Essentially, if you have a LTL int for an array and then resize the array to be some other, new array (or just assign another array into it...), then the LTL needs to be no longer true.

THIS IS A SOUNDNESS BUG

Also thanks to Suzanne for mentioning this (in an unrelated discussion) earlier; I realized while walking home that I didn't remember implementing it...

**Comment by kelloggm** [comment]
Fixed in the UBC in 555e9d27a2adf4d96a3b416f1e268ee57bf9a471

**Comment by kelloggm** [comment]
I've thought about it, and I can't come up with a way to make the MinLen checker misbehave using this sort of thing, so I'm going to close the issue. That said, I should bring this up with Mike when I next see him.

**Comment by kelloggm** [comment]
Only fixes some part of the problem

**Comment by kelloggm** [comment]
I removed the code that hides some of the symptoms of this, because it's not a complete solution.

**Comment by kelloggm** [comment]
Mike and I discussed a potential solution to both this problem and #118. Consider the following approach:

At the declaration of every array `a`, for all integers in scope `i`:
- if `i` is a compile-time constant (via the value checker) and `a` has `@MinLen(j)`, if `j > i` then `i`'s type becomes `@LTLengthOf("a")`
- if `a` has type `@SameLen("b")` and if `i` has type `@LTLengthOf("b")`, then `i`'s type becomes `@LTLengthOf({"a", "b"})`

Any time an array `a` is reassigned, for all integers `i` in scope:
- if `i`'s type refers to `a`, remove that type (i.e. `@LTLengthOf({"a", "b"})` becomes `@LTLengthOf("b")`)
- if `i` has a user-written type that refers to `a`, report an error at the reassignment.
- treat `a` as a new array declaration by following the steps above

The process described above can also be improved by adding a special case to allow the initializer of `a` when `a` is reassigned to use information about `a`'s old type (i.e. if the reassignment is `a = new int[a.length + 1]`).

The main implementation hurdles are:
- need to be able to find every integer in scope, and look up its type
- need to be able to find out whether a given integer has a user-written type or not

**Comment by smillst** [comment]
The following code would type check under this proposal, but the code still produces an index out of bounds exception.

````
class A {
  public int[] a = new int[5];
  private @LTLengthof("a") int x = a.length -1;
  int getX() {
    return a[x];
  }
}

class B {
    void test(A objA) {
      objA.a = new int[2];
      objA.getX(); // index out of bounds
   }
}
````
It's insufficient to only consider integers in scope since integers out of scope could also refer to some array. 

We already have a proposal to fix this particular bug (but not #118) in [this document](https://docs.google.com/document/d/1820riCds1WEHiOWqWPHi4Kiif7lyGtERgDY4HKvRCEY/edit).

**Comment by kelloggm** [comment]
You're right. Mike and I remembered that document yesterday but weren't sure where it was. I think we should consider actually implementing the proposal there soon; it would be really good to eliminate the only known source of unsoundness in the checker.
